pkg/gce-pd-csi-driver/controller.go

```diff
@@ -331,11 +338,149 @@ func (gceCS *GCEControllerServer) CreateVolume(ctx context.Context, req *csi.Cre
+		// Create multi-zone disk, that may have up to N disks.
+		return gceCS.createMultiZoneDisk(ctx, req, params)
+	}
+
+	// Create single device zonal or regional disk
+	return gceCS.createSingleDeviceDisk(ctx, req, params)
+}
+
+func (gceCS *GCEControllerServer) getSupportedZonesForPDType(ctx context.Context, zones []string, diskType string) ([]string, error) {
+	project := gceCS.CloudProvider.GetDefaultProject()
+	zones, err := gceCS.CloudProvider.ListCompatibleDiskTypeZones(ctx, project, zones, diskType)
+	if err != nil {
+		return nil, err
+	}
+	return zones, nil
+}
+
+func (gceCS *GCEControllerServer) getMultiZoneProvisioningZones(ctx context.Context, req *csi.CreateVolumeRequest, params common.DiskParameters) ([]string, error) {
+	top := req.GetAccessibilityRequirements()
+	if top == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "no topology specified")
+	}
+	prefZones, err := getZonesFromTopology(top.GetPreferred())
+	if err != nil {
+		return nil, fmt.Errorf("could not get zones from preferred topology: %w", err)
+	}
+	reqZones, err := getZonesFromTopology(top.GetRequisite())
+	if err != nil {
+		return nil, fmt.Errorf("could not get zones from requisite topology: %w", err)
+	}
+	prefSet := sets.NewString(prefZones...)
+	reqSet := sets.NewString(reqZones...)
+	prefAndReqSet := prefSet.Union(reqSet)
+	availableZones := prefAndReqSet.List()
+	if prefAndReqSet.Len() == 0 {
+		// If there are no specified zones, this means that there were no aggregate
+		// zones (eg: no nodes running) in the cluster
+		availableZones = gceCS.fallbackRequisiteZones
+	}
+
+	supportedZones, err := gceCS.getSupportedZonesForPDType(ctx, availableZones, params.DiskType)
+	if err != nil {
+		return nil, fmt.Errorf("could not get supported zones for disk type %v from zone list %v: %w", params.DiskType, prefAndReqSet.List(), err)
+	}
+
+	return supportedZones, nil
+}
+
+func (gceCS *GCEControllerServer) createMultiZoneDisk(ctx context.Context, req *csi.CreateVolumeRequest, params common.DiskParameters) (*csi.CreateVolumeResponse, error) {
+	// Determine the zones that are needed.
+	var err error
+
+	// For multi-zone, we either select:
+	// 1) The zones specified in requisite topology requirements
+	// 2) All zones in the region that are compatible with the selected disk type
+	zones, err := gceCS.getMultiZoneProvisioningZones(ctx, req, params)
+	if err != nil {
+		return nil, err
+	}
+
+	multiZoneVolKey := meta.ZonalKey(req.GetName(), common.MultiZoneValue)
+	volumeID, err := common.KeyToVolumeID(multiZoneVolKey, gceCS.CloudProvider.GetDefaultProject())
+	if err != nil {
+		return nil, err
+	}
+	if acquired := gceCS.volumeLocks.TryAcquire(volumeID); !acquired {
+		return nil, status.Errorf(codes.Aborted, common.VolumeOperationAlreadyExistsFmt, volumeID)
+	}
+	defer gceCS.volumeLocks.Release(volumeID)
+
+	createDiskErrs := []error{}
+	createdDisks := make([]*gce.CloudDisk, 0, len(zones))
+	for _, zone := range zones {
```

Comment:
Is it possible on a retry that the list of zones could change?

Reply:
Yes, possible if support changes from the API side.

Right now, we're filtering down the user specified list of zones into the set of compatible zones. What do you think about using this as a check (and throwing an error if one of the provided zones is not supported?). This will cause the auto-scaling scenario and scale-from-zero scenarios to not work as well in regions where not all zones are supported for `hyperdisk-ml`. But we could also notify the user with an error message like:

`Disk type abc is not supported in zones [x y z]. If using dynamic topology awareness, please explicitly specify the set of zones`.

Reply:
I was thinking more from a k8s perspective. Maybe some nodes scaled down and that particular zone is not available anymore.

Reply:
Updated this logic to check for existing disks (with the same name/type, for all zones in the region) when determining the set of zones